### PR TITLE
14548: add actions for FHIR Transforms

### PR DESCRIPTION
**Pull request by mkalish**
This PR adds an "action" to fhir->fhir elements and implements the ability to delete a resource

Test Steps:
This is covered by an automated test, but locally you can add a new element to the simple report transform that deletes a resource 

This element will remove all observations in a bundle
```yaml
  - name: delete-observations
    resource: 'Bundle.entry.resource.ofType(Observation)'
    action: "DELETE"
```

## Changes
- Introduces an "action" attribute to fhir-to-fhir transforms
- Implements the delete action

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #14548

### CODE REVIEW

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt [hunk 1]
```diff
@@ -104,9 +106,26 @@ class FhirTransformer(
...
+                        val value = getValue(element, bundle, singleFocusResource, elementContext)
+                        if (value != null) {
+                            setBundleProperty(
+                                element.bundleProperty,
+                                value,
+                                context,
+                                bundle,
+                                singleFocusResource
+                            )
+                        }
```
Comment (jalbinson) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:120]: log when a value is null so we know a no-op set happened.

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt [hunk 2]
```diff
@@ -126,25 +145,9 @@ class FhirTransformer(
...
-                                singleFocusResource
-                            )
-                        }
-                        debugMsg += "condition: true, resourceType: ${singleFocusResource.fhirType()}, " +
-                            "value: $value"
+                    FhirTransformSchemaElementAction.DELETE -> {
+                        bundle.deleteResource(singleFocusResource, removeOrphanedDiagnosticReport = false)
                     }
-
-                    // This should never happen as the schema was validated prior to getting here
```
Comment (jalbinson) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:146]: These are the comments you read when debugging a prod issue 😆

**Review by jalbinson — APPROVED** [review]
Looks good 👍

**Comment by mkalish** [comment]
@arnejduranovic This is ready for your review

File: prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/fhirTransform/FhirTransformSchema.kt [hunk 3]
```diff
@@ -79,6 +87,52 @@ class FhirTransformSchemaElement(
     valueSet = valueSet,
     debug = debug
 ) {
+
+    // Secondary constructor to Jackson can instantiate elements with an action
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/translation/hl7/schema/fhirTransform/FhirTransformSchema.kt:91]:
```suggestion
    // Secondary constructor so Jackson can instantiate elements with an action
```
?

File: prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt [hunk 4]
```diff
@@ -274,8 +274,13 @@ fun Bundle.deleteResource(resource: Base) {
...
     cleanUpReferences()
 
-    // clean up empty Diagnostic Reports and references to them
-    cleanUpEmptyDiagnosticReports()
+    // The original use case of this function was just to remove Observations from a bundle
+    // but has since expanded so this behavior is opt-in
+    if (removeOrphanedDiagnosticReport) {
+        // clean up empty Diagnostic Reports and references to them
+        cleanUpEmptyDiagnosticReports()
+    }
```
Comment (arnejduranovic) [inline prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt:282]:
```suggestion
    // TODO: Remove as part of https://github.com/CDCgov/prime-reportstream/issues/14568
    if (removeOrphanedDiagnosticReport) {
        // clean up empty Diagnostic Reports and references to them
        cleanUpEmptyDiagnosticReports()
    }
```

Comment (arnejduranovic) [reply prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt:282]: optional ^

**Review by arnejduranovic — APPROVED** [review]

**Pull request merged**